### PR TITLE
ArcTouch - Code Challenge(Kotlin)

### PR DESCRIPTION
**Pull request by alexandrebmm**
Pull request on the master, as required by the ArcTouch selection process

### CODE REVIEW

File: app/src/main/java/com/arctouch/codechallenge/activity/home/helper/MovieAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+                if (!isLoading && totalItemCount <= lastVisibleItem + visibleThreshold) {
+                    if (loadMore != null) {
+                        loadMore!!.onLoadMore()
+                    }
+                    isLoading = true
+                }
+            }
+        })
+    }
```
Comment (arctouch-leonardocasasanta) [inline app/src/main/java/com/arctouch/codechallenge/activity/home/helper/MovieAdapter.kt]: Hmm...
Looks like your adapter is doing too much work. 
What Do you think about put this init logic on the HomeActivity and let the adapter just handle RecyclerView itens?

File: app/src/main/java/com/arctouch/codechallenge/activity/BaseAppCompatActivity.kt [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import android.support.v7.app.AppCompatActivity
+import android.view.View
+import android.widget.FrameLayout
+import com.arctouch.codechallenge.R
+import com.arctouch.codechallenge.httpModule.error.ErrorCode
+import kotlinx.android.synthetic.main.app_bar_home.*
+import kotlinx.android.synthetic.main.progress_bar.view.*
+import kotlinx.android.synthetic.main.toolbar.*
+
+abstract class BaseAppCompatActivity : AppCompatActivity() {
```
Comment (arctouch-leonardocasasanta) [inline app/src/main/java/com/arctouch/codechallenge/activity/BaseAppCompatActivity.kt]: HomeActivity inherits from BaseDrawerActivity, and BaseDrawerActivity inherits from BaseAppCompatActivity.
What are the benefits and the dangerous to have hierarchy like this?

File: app/src/main/java/com/arctouch/codechallenge/activity/BaseAppCompatActivity.kt [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            releaseUi()
+
+        } else {
+            val type = ErrorCode.get(code)
+            showSnackBar(type.resId)
+        }
+        releaseUi()
+    }
+
+    protected fun blockUi() {
```
Comment (arctouch-leonardocasasanta) [inline app/src/main/java/com/arctouch/codechallenge/activity/BaseAppCompatActivity.kt]: Why this method is necessary?

**Comment by arctouch-leonardocasasanta** [comment]
Hi Alexandre!
My name is Leonardo Casasanta and I work as an Android Developer at Arctouch.
I will review your code, for now on you don't have to change your PR.
As you noticed, I ask some questions in your PR. I just want to understand your thoughts.
Please take your time to answer it.
Good luck!

**Comment by alexandrebmm** [comment]
> Hi Alexandre!
> My name is Leonardo Casasanta and I work as an Android Developer at Arctouch.
> I will review your code, for now on you don't have to change your PR.
> As you noticed, I ask some questions in your PR. I just want to understand your thoughts.
> Please take your time to answer it.
> Good luck!

Hi, Leonardo! Hope you're well.
So, I thank you, appreciate your time and valuable tips and questions. I will answer them using the 'reply' option of each one.

As you know, I had 2 days to implement, test and submit this solution. There are naturally improvements to do.
So, let's go. :)

File: app/src/main/java/com/arctouch/codechallenge/activity/BaseAppCompatActivity.kt (hunk 3 above)
Comment (alexandrebmm) [reply app/src/main/java/com/arctouch/codechallenge/activity/BaseAppCompatActivity.kt]: I should rename this function to `showLoading()`, and the other one (releaseUi) to `hideLoading()`. Probably these names would be more readable.

I use this function and its functionality to **avoid** 2 situations:
1. always insert a `ProgressBar` view in each Activity XML-layout file, and
2. Change the visibility of a whole layout and also its `ProgressBar` for every activity when necessary.

Through this implemented approach, I just need to concern about build the XML-layout and the activity properly, and nothing else. A recurrent behavior is implemented just once and becomes available for every activity.
In short, in this way I believe it is easier and more useful.

In addition, there is no possibility in missing a ProgressBar view on XML-layout and it is not needed handle visibilities across the app. If necessary to show a loading on a activity, calling this function will add dynamically a ProgressBar to the present activity, no matter which one. And the other one will remove it.

File: app/src/main/java/com/arctouch/codechallenge/activity/home/helper/MovieAdapter.kt (hunk 1 above)
Comment (alexandrebmm) [reply app/src/main/java/com/arctouch/codechallenge/activity/home/helper/MovieAdapter.kt]: Oh, of course. I agree with you. Maybe we can refactor this class and move some functions from this adapter to the `HomeActivity`, but also to the `HomePresenterImpl`, once MVP has on the View just elements to refresh directly the UI components, and the processing must be in the Presenter.

In `HomeActivity` we can add on our RecyclerView the `RecyclerView.OnScrollListener`, and other minor functions to refresh the layout, and in the `HomePresenterImpl` would be interesting to have the onLoadMore logic implemented, to decide if should request the next page. Surely it would be better

Comment (arctouch-leonardocasasanta) [reply app/src/main/java/com/arctouch/codechallenge/activity/home/helper/MovieAdapter.kt]: Nice!

File: app/src/main/java/com/arctouch/codechallenge/activity/BaseAppCompatActivity.kt (hunk 3 above)
Comment (arctouch-leonardocasasanta) [reply app/src/main/java/com/arctouch/codechallenge/activity/BaseAppCompatActivity.kt]: Oh, I see. That's smart! 
I'm just concern in terms of performance. 
Anyway, thats nice.

File: app/src/main/java/com/arctouch/codechallenge/activity/BaseAppCompatActivity.kt (hunk 2 above)
Comment (alexandrebmm) [reply app/src/main/java/com/arctouch/codechallenge/activity/BaseAppCompatActivity.kt]: Yes, exactly. Multi-level inheritance must be properly handled or using this way can lead us to a messed coding like the anti-pattern Spaguetti, what difficult the maintainance deeply.

But in this case, what happens in fact is:
1. `BaseAppCompatActivity` is a general kind of activity that every activity in this application must extend from. As you can see, it is an abstract class and has the default behavior for the application's activities, like show a loading, errors treatment and set a layout using an abstract method. `MovieDetailsActivity` extends from it directly.

2. `BaseDrawerActivity` is also abstract and is a specific kind of activity that each activity reachable from the main menu (`NavDrawer`) must extend, guaranteeing the same behavior and avoiding repetitive code for this type of activity (like onNavigationItemSelected, onBackPressed and so on).
For example, if we decide to evolve this app adding a new activity reachable from the main menu, now is pretty easier: extending from it and doing the changes at the parent will take effect in all others activities present in this main menu, and each activity just needs to implement what is properly from it. If you take a look on `AboutActivity`, you'll realize that it is really quite simple, and have just what is strictely necessary. The `HomeActivity` extends from `BaseDrawerActivity` as well. Easier maintainance on the main menu, adding, removing or updating its items.

And this inheritance makes sense, once the code reuse obbeys the principle "is a kind of", in each level. And other abstract activities aren't necessary anymore, because every other new activity out of the main menu can extend from `BaseAppCompatActivity` directly, including or not fragments, for example.

These are some of benefits:
- helping avoid repetitive code, implementing the DRY principle (Don’t Repeat Yourself)
- It extends their parent, inheriting its features
- Improves code reusability.

File: app/src/main/java/com/arctouch/codechallenge/activity/BaseAppCompatActivity.kt (hunk 3 above)
Comment (alexandrebmm) [reply app/src/main/java/com/arctouch/codechallenge/activity/BaseAppCompatActivity.kt]: Thanks! I agree with you about the performance. But for me the cost benefit is better. I have no others concerns along the all over application. Can you imagine an application with dozens of activities, http-requests and so on? I think the maintainance using this approach is better. Like I said: the cost benefit is higher.
